### PR TITLE
Upgrade CoreDNS-Operator to v2

### PR DESCRIPTION
**Pull request by rajansandeep**
The CoreDNS operator is built using Kubebuilder v2.3 and [a bug fix which fixes the plugin addon template](https://github.com/kubernetes-sigs/kubebuilder/pull/1413).

Edit: Since then v2.3.1 has been released which includes the bug fix

This replaces the existing v1 operator
(If the changes here are confusing due to file overlap, or we prefer to keep v1, I can make these changes to a different directory)

I have also updated the CoreDNS version to 1.6.7 (Current latest)

TODO: 
 - [x] Boilerplate
 - [x] Operator runs outside cluster
 - [x] Operator runs in-cluster
 - [x] Support CoreDNS upgrade path
 - [x] Support Kustomize for config hash
 - [x] Golden-tests

Fixes: https://github.com/kubernetes-sigs/cluster-addons/issues/24

### CODE REVIEW

**Comment by rajansandeep** [comment]
The Kustomize support complies using patch from https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/pull/91

**Comment by rajansandeep** [comment]
/assign @johnsonj 
PTAL when possible, thank you!

**Comment by johnsonj** [comment]
wohoo! will do- thank you @rajansandeep !

File: coredns/Makefile [hunk 1]
```diff
@@ -1,66 +1,85 @@
 
 # Image URL to use all building/pushing image targets
-IMG ?= ${USER}/coredns-operator:latest
+IMG ?= rajansandeep/coredns-operator:latest
```
Comment (johnsonj) [inline coredns/Makefile:3]: nit: revert back to ${USER}

**Comment by rajansandeep** [comment]
> I'd suggest in this change (or another) to beef up the README since this thing is no toy now!

@johnsonj, Definitely! I think the README should be focusing on "How to use/install" the operator than "How to create". 

I'll make those changes in a separate PR.

**Pull request merged**